### ISSUE TITLE
[Proposal]: Add let expression & let binding

### ISSUE THREAD

**Issue by bigopon** [aurelia/templating-binding#115]
**I'm submitting a proposal**

This is my attempt of trying to continue the work from [Jeremy POC](https://github.com/aurelia/templating-resources/issues/171#issuecomment-166159327), where Jeremy described that creating a let custom element is not ideal as multiple binding instances will be created for the same prop, and also gave a solution of 

* **SyntaxInterpreter**: Adjust bind method of this class. There are 2 cases of `bind`: let element & not let element:

```js
SyntaxInterpreter.prototype.bind = 

  /**
   * @param {ViewResources} resources
   * @param {Element} element
   * @param {{ attrName: string, attrValue: string, defaultBindingMode: number }} info
   * @param {BehaviorInstruction} existingInstruction
   */
  function(resources, element, info, existingInstruction, context) {
    let instruction = existingInstruction || BehaviorInstruction.attribute(info.attrName);

    let attributeInstruction;
    let isLet = element.localName === 'let';
    if (isLet) {
      attributeInstruction = new LetExpression(
        this.observerLocator,
        camelCase(info.attrName),
        this.parser.parse(info.attrValue),
        resources.lookupFunctions
      );
    } else {
      attributeInstruction = new BindingExpression(
        this.observerLocator,
        this.attributeMap.map(element.tagName, info.attrName),
        this.parser.parse(info.attrValue),
        info.defaultBindingMode || this.determineDefaultBindingMode(element, info.attrName, context),
        resources.lookupFunctions
      );
    }
    instruction.attributes[info.attrName] = attributeInstruction;

    return instruction;
  }
```

* **LetExpression**: This expression will create `Let` binding, that is, compared to normal bindings, different: instead of getting the target from `constructor`, it gets the target in `bind`. ([Jeremy original comment](https://github.com/aurelia/templating-resources/issues/171#issuecomment-166210281))

```js
  bind(source) {
    if (this.isBound) {
      if (this.originalSource === source) {
        return;
      }
      this.unbind();
    }
    this.isBound = true;
    this.setSource(source);
    
    if (this.sourceExpression.bind) {
      this.sourceExpression.bind(this, this.source, this.lookupFunctions);
    }
    
    enqueueBindingConnect(this);
  }

  setSource(source) {
    this.originalSource = source;
    this.source = createOverrideContext(
      source.overrideContext.bindingContext,
      source.overrideContext.parentOverrideContext
    );
    this.target = this.source.bindingContext;
  }
```

The rest is the same with everywhere else in Aurelia.

Note: 
  * This just addresses half of the concerns (creating unnecessary bindings), the other half is to skill inspecting the `let` element in `ViewCompiler` for html behavior.
  * Usage together with value converter needs to be documented as it uses `toView` instead of `fromView`, which imo, are both confusing in case of `let`
  * [Demo](https://gist.run/?id=e2fcc0d2c7e34a0c1adf806bbc5836be)

cc @EisenbergEffect @jdanyow

**Comment by jdanyow** [comment]
Looking good so far, before implementation work starts we should nail down a few details. There are a few big design decisions remaining that will have a big impact on how this is done.

1. What do we want the API to look like?
    ```html
    <let foo="bar.baz | capitalize"></let>
    ```
    or
    ```html
    <let foo.bind="bar.baz | capitalize"></let>
    ```
    Do we need the `.bind` or is it implied by the `<let>` usage? I was thinking we should use `.bind`/`.one-time`/`one-way` to be consistent with everything else.

2. Interpolation permitted? `<let foo="${something}">` (I was thinking yes)

2. The binding modes should be locked down to `one-way` and `one-time`.

3. Should we remove the `<let>` element from template so that it's never rendered. Like `<require>`. Where would this happen? In the ViewCompiler? `<require>` elements are removed in [loader/TemplateRegistryEntry](https://github.com/aurelia/loader/blob/master/src/template-registry-entry.js#L82)

4. Let bindings need to be bound before regular bindings. Otherwise bindings using the "variables" declared by the `<let>` element will need to be updated when the `<let>`'s bindings are eventually bound. This would be bad for perf. Additionally, the "variables" need to be defined otherwise the binding logic will search up the scopes for something with that name. Here's where the binding order of bindings, controllers, etc is defined: [`View.bind`](https://github.com/aurelia/templating/blob/master/src/view.js#L144).

**Comment by EisenbergEffect** [comment]
1. Let's require `.bind` for consistency reasons.
2. Let's allow interpolation, for consistency as well.
3. Definitely.
4. We should remove this if possible, but I don't think we can do it in the registry. We probably need to do it as part of the compilation step.
5. This seems like the tricky part. We may need to introduce a new set of instructions for variables that are stored in the view factory and processed during the create call before the normal instructions. I'm open to ideas here.

**Comment by bigopon** [comment]
Ive thought about removing, the `let`. In order to avoid performance impact while walking the tree, we should remove it right after finished compilation **only**, or leave that job to the view factory in its instanciate process ?

About binding order, I've never thought of it. Do we have anything that can instruct bindings bound orders ?

**Comment by EisenbergEffect** [comment]
We can remove the let during the compilation process, potentially right after the node is processed. We can hardcode this into the compiler, if we want. We don't want to do it during view instantiation since that would need to be done for every instantiate, when it could be done once for all.

All instructions (bindings, elements, etc) are processed in depth-first order as returned from a DOM query selector. We don't want to change that. We might just need to add something special for variables.

**Comment by bigopon** [comment]
What about [this line](https://github.com/aurelia/templating/blob/master/src/view-compiler.js#L112)

```js
let factory = new ViewFactory(content, instructions, resources);
```

It's in `compile` method of view compiler, we can insert a remove `let` there without any issue, and it will be done only once, also ensure depth-first order I think ... (not 100% sure)

**Comment by jdanyow** [comment]
@bigopon I suggest stepping through the ViewCompiler with the debugger to get a sense of how things are instantiated, first with a view that has one or two bindings, then with something more complex like a view with custom elements, custom attributes, etc. Finally, something with template parts or slots.  There's a lot to digest in there but once you're familiar with this process the rest of the internals will make more sense.  Possibly take a look at TemplateRegistryEntry first. Let us know if you have any questions.

**Comment by EisenbergEffect** [comment]
I think it can more easily and efficiently be done earlier than that when the actual nodes are being walked. Right around here: https://github.com/aurelia/templating/blob/master/src/view-compiler.js#L304 and using a similar technique as slots.

**Comment by bigopon** [comment]
That was such a bad comment, what I fully meant was we just skip everything if we encounter the let (Similar to what Rob said, except that we leave it in the template and queryAll / remove later):

```js
if (tagName === 'let') {
  // inspect all of its attribute and bail
  // Because we exit the node after extracted its attributes, we can easily remove it after the whole compilation process has been done
}
```

But I agree that I need to setup a debugger. Been wanting to do that, just those almost hundred lines of if / else scared me every time.

**Comment by bigopon** [comment]
About binding order, I just looked at the demo again and think there is no right order.

Usual case is normal bindings depend on some of let bindings, and let bindings depend on some normal bindings. Either way, some thing always need to go in first. So I think unless we do very intelligent dependency tracking there, there is no way to tell which is the right order

Can @jdanyow @EisenbergEffect give some thoughts ?

Edit: Actually I can continue with the assumption that let bindings always depend on normal bindings, and normal bindings may / may not depend on let bindings, So let should always go after, it's safer.

**Comment by jdanyow** [comment]
I think let element bindings are highest priority for the reasons I mentioned in https://github.com/aurelia/templating-binding/issues/115#issuecomment-327371723

If we're targeting this use case: https://github.com/aurelia/templating-resources/issues/171#issue-123178218, this holds true.

Assuming folks don't position the let element after elements using the "variables" it creates, the bindings may already be in the correct order.

**Comment by EisenbergEffect** [comment]
If we want to keep the dom position, then we may need to leave the let element in, since the view factories create method processes based on dom order.

**Comment by bigopon** [comment]
tl;dr

I have brought all desired behaviors in #issuecomment-327371723. A few lines of key things and remaining challenge (I couldn't solve)

#### Extracting let expressions

```js
let instructions = { letExpressions: [] }; // instead of let instruction = {};
// start compiling

// When hit a let element:
else if (tagName === 'let') {
  bindingLanguage.createLetExpressions(
    resources,
    node,
    instructions.letExpressions // this is from above. all instruction in let element will be here
  );
  nextSib = node.nextSibling;
  DOM.removeNode(node);
  return nextSib;
}
```

The remaining problem I couldn't solve is what to do when I hit something like this:

```html
<template>

  <my-custom-element>
    <let target.bind='source'></let>
  </my-custom-element>
</template>
```

 It works fine in template controller context but this is quite confusing to me and will it have any affect on future, when we want to allow custom element to be template controller ?

#### binding mode

In normal binding, `one-time` make sense because it has only 1 source, thus the source updates its subscribers only once. In the let context, it's source, but should be understood as target, and thus cannot really work with `one-time` behavior, imo.

--------------------------------------------
Original comment:

I have finished the instructions above, but it's a lot of patching to be able to demo on Gist now, so i'll summarize them below:

  1. **View Compiler**
    Adjust the `_compileElement` so when it hit the `let` element, it will extract the attribute into let binding expressions, create TargetInstruction for the let element, and use the element itself instead of creating another element

<details>
```js
    function markLetTarget(resources, letElement, instructions, expressions) {
      let auTargetID = makeIntoInstructionTarget(letElement);
      let instruction = TargetInstruction.letExpression(expressions);
      instructions[auTargetID] = instruction;
    }

    // ....

    else if (tagName === 'let') {
      letExpressions = bindingLanguage.createLetExpressions(resources, node);
      markLetTarget(resources, node, instructions, letExpressions);
      return node.nextSibling;
    }
```
</details>

  2. **Target Instruction**
    Add boolean flag in constructor `isLetInstruction`, to guide `applyInstructions` in `ViewFactory.prototype.create` to just get the target expressions and **`unshift`** it to the start of binding list. (@jdanyow because of this I don't think we need to sort the bindings in `View.prototype.create` ? )

<details>
```js
    // In apply instructions
    if (instruction.isLetInstruction) {
      let letCommentAnchor = DOM.createComment('let');
      DOM.replaceNode(letCommentAnchor, element);

      for (i = 0, ii = expressions.length; ii > i; ++i) {
        bindings.unshift(expressions[i].createBinding());
      }
      return;
    }
```
</details>

  3. **BindingLanguage**
    Add `createLetExpressions`, the implementation in `TemplateBindingLanguage`:

<details>

```js
      createLetExpressions(resources, letElement) {
        let attributes = letElement.attributes;
        /**@type {Attr} */
        let attr;
        /**@type {LetExpression[]} */
        let expressions = [];
        /**@type {string[]} */
        let parts;
        let attrName;
        let attrValue;
        let command;
        let mode;
        // debugger;
        for (let i = 0, ii = attributes.length; ii > i; ++i) {
          attr = attributes[i];
          attrName = attr.name;
          attrValue = attr.nodeValue;
          parts = attrName.split('.');
          if (parts.length === 2) {
            command = parts[1];
            if (validLetCommands.indexOf(command) === -1) {
              LogManager.getLogger('templating-binding')
                .warn(`Detected invalid let command. Expected one of ${validLetCommands}, given "${command}"`);
              continue;
            }
            mode = command === 'bind' ? bindingMode.oneWay : bindingMode[camelCase(command)];
            expressions.push(new LetExpression(
              this.observerLocator,
              camelCase(parts[0]),
              this.parser.parse(attrValue),
              mode,
              resources.lookupFunctions
            ));
          } else {
            let parts = this.parseInterpolation(resources, attrValue);
            if (parts === null) {
              LogManager.getLogger('templating-binding')
                .warn(`Detected string literal in let bindings. Did you mean "${ attrName }.bind=${ attrValue }" or "${ attrName }=\${${ attrValue }}" ?`);
              continue;
            }
          }
        }
        return expressions;
      }
```

</details>

I'm implementing `LetInterpolationExpression`.

Can I have some suggestions @jdanyow @EisenbergEffect  ?

**Edit 1**: <details>Actually I don't think I need to replace the let element with anything, can be removed from the DOM safely, am I correct ?</details>
**Edit 2**: <details>I think removing is safe, if so, after compiling a factory, maybe I can just go a head and loop over the instructions, find the `TargetInstruction` with `isLetBinding === true`, merge their expressions back into the factory's new property `letExpressions` to create the binding before applying the instruction. Doing it this way we don't need to remove the let element over and over. 

```js
    // In the compile method, after finished compiling the view, inspect anchor is container, we merge the let instructions

    let letExpressions = [];
    let letInstructionIds= [];
    for (let instructionId in instructions) {
      let instruction = instructions[instructionId];
      let isLetInstruction = instruction.isLetInstruction;
      if (isLetInstruction) {
        letInstructionIds.push(instructionId);
      }
      if (isLetInstruction && instruction.expressions.length !== 0) {
        letExpressions = letExpressions.concat(instruction.expressions);
      }
    }

    for (let i = 0; ii = letInstructionIds.length; ii > i; ++i) {
      delete instructions[letInstructionIds[i]];
    }

    let factory = new ViewFactory(content, instructions, resources, letExpressions);
```

</details>

**Edit 3**: 

<details>From edit 2, I think we can start compiling in view compiler with 

```js
let instructions = { letExpressions: [] };
// instead of
let instruction = {};
```

  Then whenever compiler finds a let element, extract all expression to the `letExpressions` and remove the element.</details>

**Edit 4**: 

<details>I've try the above approach and it seems better. Just need a little tweak  [at this line in `ViewFactory.prototype.create`](https://github.com/aurelia/templating/blob/master/src/view-factory.js#L433)

```js
// old
let binding = [];
// new
let bindings = this.letExpressions.map(e => e.createBinding());
```

</details>

**Edit 5**: 

<details>In the following case

```html
<template>

  <my-custom-element>
    <let target.bind='source'></let>
  </my-custom-element>
</template>
```

What should it do in this case ? It works fine in template controller context but this is quite confusing to me and will it have any affect on future, where we want to allow custom element to be template controller ?</details>

**Comment by bigopon** [comment]
I think for this scenario

```html
<!-- app.html -->
<template>

  <my-custom-element>
    <let target.bind='source'></let>
  </my-custom-element>
</template>
```

It can simulate the "scope hoisting" feature of JS, so the let belongs to the `app` element. But then we need to make the documentation clear on which situation it wont be hoisted.

@EisenbergEffect @jdanyow What do you think ?

**Comment by bigopon** [comment]
I have finished let string interpolation. It will leverage existing classes: `InterpolationBinding` and `ChildInterpolationBinding`.

[Update Demo](https://gist.run/?id=e2fcc0d2c7e34a0c1adf806bbc5836be)

Total changes are:

  * `BindingLanguage` new API & `TemplatingBindingLanguage` implementation
```ts
  class BindingLanguage {
    createLetExpressions(
      resources: ViewResources,
      node: /*let*/Element,
      existingLetExpressions: LetExpression[]
    ): LetExpression[]
  }
```

  * 4 new classes: `LetExpression`, `LetBinding`, `LetInterpolationBindingExpression`, `LetInterpolationBinding`

  * `ViewCompiler.prototype.compile` [line 97](https://github.com/aurelia/templating/blob/master/src/view-compiler.js#L97)

```js
  // change from
  // let instructions = {};
  // to
  let instructions = { letExpressions: [] };
```

  * `ViewCompiler.prototype._compileElement`:

```js
    ...
      else if (tagName === 'let') {
        bindingLanguage.createLetExpressions(
          resources,
          node,
          instructions.letExpressions // this is from above. all instruction in let element will be here
        );
      nextSib = node.nextSibling;
      DOM.removeNode(node);
      return nextSib;
    }
    ...
```

  * `ViewFactory.prototype.create` [line 433](https://github.com/aurelia/templating/blob/master/src/view-factory.js#L433). This align with @jdanyow 's concerns that let bindings should come in first.

```js
  // change from
  // let bindings = []
  // to
  let bindings = this.instructions.letExpressions.map(e => e.createBinding());
```

@EisenbergEffect @jdanyow

**Comment by jdanyow** [comment]
This is looking great. Nice work. Time for PRs? Minor suggestion- use loop instead of `.map` for perf :fire:

**Comment by bigopon** [comment]
😄 

I have another [updated demo here](https://gist.run/?id=80d070452f6c5afbe9173571422da317)

Please have a look at the last section, common confusing demo. I basically wanted to make it clear that `let` behavior will not be altered at all, when put in side an `if`, or custom element, or custom element with slot. With default implementation, only `repeat` will have in on the on-the-fly scope. Do you feel anything wrong ? If not then that's great. I just wanna make sure 😄 

About PRs, I think i will separate new classes to:

  * `LetExpression`, `LetBinding` to `binding`
  * `LetInterpolationBindingExpression`, `LetInterpolationBinding` to `templating-binding`

Is is good or everything should just stay here ?

@jdanyow

**Comment by bigopon** [comment]
Resolved by #117
